simavr/sim/avr_uart.c

```diff
@@ -239,8 +239,27 @@ avr_uart_baud_write(
 		void * param)
 {
 	avr_uart_t * p = (avr_uart_t *)param;
+	if (addr == p->ubrrh.reg) {
+		if (p->ubrrh.reg == p->r_ucsrc) {
+			// UBRRH and UCSRC registers can share the same I/O location.
+			// URSEL bit decides which register gets written. If it's UBRRH,
+			// store to shadow register, else to regular register
+			if ((v & (1 << 7)) == 0) { // URSEL
+				avr_regbit_setto_shadow(avr, p->ubrrh, v, &p->ubrrh_shadow);
+			} else {
+				avr_core_watch_write(avr, addr, v);
+			}
+		} else {
+			// If UBRRH have different I/O locations, update shadow register and
+			// regular register
+			avr_core_watch_write(avr, addr, v);
+			avr_regbit_setto_shadow(avr, p->ubrrh, v, &p->ubrrh_shadow);
+		}
+		return;
+	}
```

Comment:
I think that sort of behaviour should be encoded in the core declaration, for example a new field with 'ursel' would both allow feature testing and encode the feature in a generic way. here it's all hard coded.
Also for that sort of feature, I don't /think/ we really need the new _shadow accessors for regbits, it's a simple enough behaviour that is just visible in the set/get for this feature, AFAIK...

Personally, I would introduce a more generic behaviour where the two (or more) values are stored part of a regbit with a discriminant, instead of having to carry around a pointer to an uint8_t...